xbmc/platform/darwin/ios-common/DarwinEmbedNowPlayingInfoManager.mm

```diff
@@ -0,0 +1,149 @@
+    dict[MPMediaItemPropertyArtist] = [artists componentsJoinedByString:@" "];
+  if (NSNumber* track = item[@"track"])
+    dict[MPMediaItemPropertyAlbumTrackNumber] = track;
+  if (NSNumber* duration = item[@"duration"])
+    dict[MPMediaItemPropertyPlaybackDuration] = duration;
+  NSArray* genres = item[@"genre"];
+  if (genres && genres.count > 0)
+    dict[MPMediaItemPropertyGenre] = [genres componentsJoinedByString:@" "];
+
+  NSString* thumb = item[@"thumb"];
+  if (thumb && thumb.length > 0)
+  {
+    if (auto image = [UIImage imageWithContentsOfFile:thumb])
+    {
+      if (auto mArt =
+          [[MPMediaItemArtwork alloc] initWithBoundsSize:image.size
+                                          requestHandler:^UIImage* _Nonnull(CGSize aSize) {
+                                            return image;}])
+      {
+        dict[MPMediaItemPropertyArtwork] = mArt;
+      }
+    }
+  }
+
+  if (NSNumber* elapsed = item[@"elapsed"])
+    dict[MPNowPlayingInfoPropertyElapsedPlaybackTime] = elapsed;
+  if (NSNumber* speed = item[@"speed"])
+    dict[MPNowPlayingInfoPropertyPlaybackRate] = speed;
+  if (NSNumber* current = item[@"current"])
+    dict[MPNowPlayingInfoPropertyPlaybackQueueIndex] = current;
+  if (NSNumber* total = item[@"total"])
+    dict[MPNowPlayingInfoPropertyPlaybackQueueCount] = total;
+
+  /*! @Todo additional properties?
+   other properities can be set:
+   MPMediaItemPropertyAlbumTrackCount
+   MPMediaItemPropertyComposer
+   MPMediaItemPropertyDiscCount
+   MPMediaItemPropertyDiscNumber
+   MPMediaItemPropertyPersistentID
+   Additional metadata properties:
+   MPNowPlayingInfoPropertyChapterNumber;
+   MPNowPlayingInfoPropertyChapterCount;
+   */
+
+  [self setNowPlayingInfo:dict];
+
+  self.playbackState = DARWINEMBED_PLAYBACK_PLAYING;
+
+  [g_xbmcController disableNetworkAutoSuspend];
+}
+
+- (void)OnSpeedChanged:(NSDictionary*)item
+{
+  NSMutableDictionary* info = [self.nowPlayingInfo mutableCopy];
+  if (NSNumber* elapsed = item[@"elapsed"])
+    info[MPNowPlayingInfoPropertyElapsedPlaybackTime] = elapsed;
+  if (NSNumber* speed = item[@"speed"])
+    info[MPNowPlayingInfoPropertyPlaybackRate] = speed;
+
+  [self setNowPlayingInfo:info];
+}
+
+- (void)onPause:(NSDictionary*)item
+{
+  self.playbackState = DARWINEMBED_PLAYBACK_PAUSED;
+
+  // schedule set network auto suspend state for save power if idle.
+  [g_xbmcController rescheduleNetworkAutoSuspend];
+}
+
+- (void)onStop:(NSDictionary*)item
+{
+  [self setNowPlayingInfo:nil];
+
+  self.playbackState = DARWINEMBED_PLAYBACK_STOPPED;
+
+  // delay set network auto suspend state in case we are switching playing item.
+  [g_xbmcController rescheduleNetworkAutoSuspend];
+}
+
+- (void)observeValueForKeyPath:(NSString*)keyPath
+                      ofObject:(id)object
+                        change:(NSDictionary*)change
+                       context:(void*)context
+{
+  if (![keyPath isEqualToString:NSStringFromSelector(@selector(nowPlayingInfo))])
```

Comment:
if you write `!` here, then next statement should be `return` :) just drop the `!`